core/shared/src/main/scala/fs2/Compiler.scala

```diff
@@ -114,22 +114,47 @@ object Compiler extends CompilerLowPriority {
         .unsafeRunSync()
   }
 
-  sealed trait Target[F[_]] extends Resource.Bracket[F] {
+  sealed trait Target[F[_]] extends MonadCancelThrow[F] {
     def ref[A](a: A): F[Ref[F, A]]
-    def uncancelable[A](f: Poll[F] => F[A]): F[A]
     private[fs2] def interruptContext(root: Token): Option[F[InterruptContext[F]]]
   }
 
   private[fs2] trait TargetLowPriority {
-    implicit def forSync[F[_]: Sync]: Target[F] = new SyncTarget
+    protected abstract class MonadErrorTarget[F[_]](implicit F: MonadError[F, Throwable])
+        extends Target[F] {
+      def pure[A](a: A): F[A] = F.pure(a)
+      def handleErrorWith[A](fa: F[A])(f: Throwable => F[A]): F[A] = F.handleErrorWith(fa)(f)
+      def raiseError[A](e: Throwable): F[A] = F.raiseError(e)
+      def flatMap[A, B](fa: F[A])(f: A => F[B]): F[B] = F.flatMap(fa)(f)
+      def tailRecM[A, B](a: A)(f: A => F[Either[A, B]]): F[B] = F.tailRecM(a)(f)
+    }
+
+    def uncancelable[F[_]](implicit F: Sync[F]): Target[F] = new MonadErrorTarget[F]()(F) {
```

Comment:
I left this public despite providing an unlawful instance of `MonadCancelThrow` in case anyone wants to compile a stream to `Coeval` or another `Sync[F]` that doesn't support cancelation. Kind of a hack.